modules/apps/27-interchain-accounts/types/packet.go

```diff
@@ -48,6 +54,114 @@ func (iapd InterchainAccountPacketData) GetBytes() []byte {
 	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(&iapd))
 }
 
+/*
+
+ADR-8 CallbackPacketData implementation
+
+InterchainAccountPacketData implements CallbackPacketData interface. This will allow middlewares targeting specific VMs
+to retrieve the desired callback address for the ICA packet on the source chain. Destination callback addresses are not
+supported for ICS 27.
+
+The Memo is used to set the desired callback addresses.
+
+The Memo format is defined like so:
+
+```json
+{
+	// ... other memo fields we don't care about
+	"src_callback": {
+		"address": {stringForContractAddress},
+
+		// optional fields
+		"gas_limit": {stringForGasLimit},
+	}
+}
+```
+
+*/
+
+// GetSourceCallbackAddress returns the source callback address if it is specified in the packet data memo.
+// If no callback address is specified or is improperly formatted, an empty string is returned.
+//
+// The memo is expected to contain the source callback address in the following format:
+// { "src_callback": { "address": {stringCallbackAddress}}
+//
+// ADR-8 middleware should callback on the returned address if it is a PacketActor
+// (i.e. smart contract that accepts IBC callbacks).
+func (iapd InterchainAccountPacketData) GetSourceCallbackAddress() string {
+	callbackData := iapd.getCallbackData("src_callback")
+	if callbackData == nil {
+		return ""
+	}
+
+	callbackAddress, ok := callbackData["address"].(string)
+	if !ok {
+		return ""
+	}
+
+	return callbackAddress
+}
+
+// GetDestCallbackAddress returns an empty string. Destination callback addresses
+// are not supported for ICS 27. This feature is natively supported by
+// interchain accounts host submodule transaction execution.
+func (iapd InterchainAccountPacketData) GetDestCallbackAddress() string {
+	return ""
+}
+
+// GetSourceUserDefinedGasLimit returns the custom gas limit provided for source callbacks
+// if it is specified in the packet data memo.
+// If no gas limit is specified or the gas limit is improperly formatted, 0 is returned.
+//
+// The memo is expected to specify the user defined gas limit in the following format:
+// { "src_callback": { ... , "gas_limit": {stringForGasLimit} }
+func (iapd InterchainAccountPacketData) GetSourceUserDefinedGasLimit() uint64 {
+	callbackData := iapd.getCallbackData("src_callback")
```

Comment:
Sounds good. Where should these constants live?
- In callbacks module so that they can be imported here
- In the same `types/` directory as `packet.go` (some duplication will occur here)
- Other ideas?